### PR TITLE
drivers: watchdog: mtl: Add driver for the Synopsys DesignWare watchdog

### PR DESCRIPTION
**Pull request by softwarecki**
This PR adds a base version of a driver for the Synopsys DesignWare watchdog IP. Also added watchdog device instances to the device tree of mtl platform.

### CODE REVIEW

**Review by marcinszkudlinski — DISMISSED** [review]
looks fine

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi [hunk 1]
```diff
@@ -424,6 +424,30 @@
 			interrupt-parent = <&ace_intc>;
 		};
 
+		watchdog0: watchdog0@78300 {
```
Comment (galak) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:427]: should just be watchdog0: watchdog@78300

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi [hunk 2]
```diff
@@ -424,6 +424,30 @@
...
+			compatible = "snps,designware-watchdog";
+			reg = <0x78300 0x100>;
+			clock-frequency = <32768>;
+			reset-pulse-length = <2>;
+			status = "okay";
+		};
+
+		watchdog1: watchdog1@78400 {
```
Comment (galak) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:435]: should just be watchdog1: watchdog@78400

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi [hunk 3]
```diff
@@ -424,6 +424,30 @@
...
+			compatible = "snps,designware-watchdog";
+			reg = <0x78400 0x100>;
+			clock-frequency = <32768>;
+			reset-pulse-length = <2>;
+			status = "okay";
+		};
+
+		watchdog2: watchdog2@78500 {
```
Comment (galak) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:443]: should just be watchdog2: watchdog@78500

**Review by galak — DISMISSED** [review]
Some minor changes to dtsi required.

Not sure if we should match the linux binding for this:

https://www.kernel.org/doc/Documentation/devicetree/bindings/watchdog/snps%2Cdw-wdt.yaml

File: dts/bindings/watchdog/snps,designware-watchdog.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    reg:
+      required: true
+
+    interrupts:
+
+    clocks:
+
+    clock-frequency:
+      type: int
+      description: Clock frequency used by counter
```
Comment (marcinszkudlinski) [inline dts/bindings/watchdog/snps,designware-watchdog.yaml:17]: unit? Hz, MHz ?

File: dts/bindings/watchdog/snps,designware-watchdog.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  clock-frequency:
+    type: int
+    description: |
+      Clock frequency used by counter in Hz. You can specify a frequency here or specify a clock
+      using the clocks propertie.
+
+  reset-pulse-length:
+    type: int
+    description: Duration of the reset pulse in clock cycles (2/4/8/16/32/64/128/256)
```
Comment (galak) [inline dts/bindings/watchdog/snps,designware-watchdog.yaml:23]: Might make this an enum if there is a fixed set of values.

**Review by galak — COMMENTED** [review]
Devicetree bits look good.  One comment about possibly using enum for 'reset-pulse-length'.

**Review by abonislawski — DISMISSED** [review]

**Review by tmleman — DISMISSED** [review]

**Review by aborisovich — CHANGES_REQUESTED** [review]
I see some potential issues here related to pausing the watchdog.
Power management for watchdog device is not implemented and device disabling function just returns -EINVAL. It looks to me that we can not stop watchdog from working?
If so, several questions come to my mind:
1) Do we need to signal watchdog during system power transition? During power transition devices used by application are probably locked and disabled one by one. I do understand that signaling a watchdog will be probably done in some separate thread by the application but signals send to watchdog always have dependencies on some system actions (otherwise just signaling would not make sense). We should probably stop watchdog timer eg. in the registered by application [pm_notifier](https://docs.zephyrproject.org/2.7.0/reference/power_management/index.html#c.pm_notifier_register). Not to mention actions like dumping a memory for context save/restore...
2) Do we need to signal watchdog timer during failed [runtime assertion](https://docs.zephyrproject.org/3.0.0/reference/kernel/other/fatal.html#runtime-assertions)?
Let's assume application had enabled CONFIG_ASSERT, some assertion failed, default `assert_post_action()` had been triggered and system spins in some kind of while(true) loop.
We can assume that application in such state has all interrupts disabled, but if the watchdog is configured to system reset?

To sum up:
it looks to me that watchdog should be stoppable in some scenarios I mentioned.
Especially the case with power transition - without watchdog device pausing we may not be able to enter sleep state and leave it successfully.

File: drivers/watchdog/wdt_dw.c [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+};
+
+static int dw_wdt_setup(const struct device *dev, uint8_t options)
+{
+	const struct dw_wdt_dev_cfg *const dev_config = dev->config;
+	struct dw_wdt_dev_data *const dev_data = dev->data;
+	uint32_t period;
+
+	if (options & WDT_OPT_PAUSE_HALTED_BY_DBG) {
+		LOG_ERR("Pausing watchdog by debugger is not supported.");
```
[the remark this answers is not in the document]
Comment (softwarecki) [reply drivers/watchdog/wdt_dw.c:50]: The watchdog hardware is independent of the CPU core and it cannot detect that core is debugged.

File: drivers/watchdog/wdt_dw.c [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+		return -ENOTSUP;
+	}
+
+	if (options & WDT_OPT_PAUSE_IN_SLEEP) {
+		LOG_ERR("Pausing watchdog in sleep is not supported.");
```
[the remark this answers is not in the document]
Comment (softwarecki) [reply drivers/watchdog/wdt_dw.c:55]: The watchdog hardware is independent of the CPU core and it cannot detect that core is in sleep state. Watchdog IP can support pause signal (this is the ip configuration option) but it is a hardware signal. It is impossible to pause the watchdog from its registers.

File: drivers/watchdog/wdt_dw.c [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+		),                                                                                 \
+		.reset_pulse_length = ilog2(DT_INST_PROP_OR(inst, reset_pulse_length, 2)) - 1,     \
+		IF_ENABLED(IS_INST_IRQ_EN(inst),                                                   \
+			(.irq_config = dw_wdt##inst##_irq_config,)                                 \
+		)                                                                                  \
+	};                                                                                         \
+                                                                                                   \
+	static struct dw_wdt_dev_data wdt_dw##inst##_data;                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, &dw_wdt_init, NULL, &wdt_dw##inst##_data,                      \
```
[the remark this answers is not in the document]
Comment (softwarecki) [reply drivers/watchdog/wdt_dw.c:235]: > device disabling function just returns -EINVAL.
Once the watchdog is enabled, it is impossible to stop it. The hardware will ignore clearing a enable bit.

**Review by marcinszkudlinski — DISMISSED** [review]

File: drivers/watchdog/wdt_dw.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (marcinszkudlinski) [reply drivers/watchdog/wdt_dw.c:55]: we need to develop a GDB script turning the WTD off when debugger is connected
Alternatively use a debug overlay

File: drivers/watchdog/wdt_dw.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (abonislawski) [reply drivers/watchdog/wdt_dw.c:50]: probably wdt will be first thing to disable for debug

[the remark this answers is not in the document]
Comment (softwarecki) [reply drivers/watchdog/wdt_dw.c:50]: Definitely yes. Fortunately, it is a signle option to disable in a config.

[the remark this answers is not in the document]
Comment (aborisovich) [reply drivers/watchdog/wdt_dw.c:50]: Ok, so we will just disable it statically building firmware configuration for debugging - got it.

**Review by aborisovich — DISMISSED** [review]
All my concerns had been addressed.

**Review by abonislawski — DISMISSED** [review]

**Review by aborisovich — DISMISSED** [review]

**Review by marcinszkudlinski — DISMISSED** [review]

File: drivers/watchdog/wdt_dw.c [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+#if WDT_DW_INTERRUPT_SUPPORT
+static void dw_wdt_isr(const struct device *dev)
+{
+	const struct dw_wdt_dev_cfg *const dev_config = dev->config;
+	struct dw_wdt_dev_data *const dev_data = dev->data;
+
+	if (dw_wdt_interrupt_status_register_get(dev_config->base)) {
+		dw_wdt_clear_interrupt(dev_config->base);
+
+		if (dev_data->callback)
```
Comment (ceolin) [inline drivers/watchdog/wdt_dw.c:186]: you need braces here

Comment (ceolin) [reply drivers/watchdog/wdt_dw.c:186]: and in other similar cases

File: drivers/watchdog/wdt_dw.c [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+			dev_data->callback(dev, 0);
+	}
+}
+#endif
+
+#define CHECK_RESET_PULSE_LENGTH_VALUE(rpl)                                                        \
```
Comment (ceolin) [inline drivers/watchdog/wdt_dw.c:192]: This macro name is pretty confuse, it is just if the pulse length value is valid right ?

File: drivers/watchdog/wdt_dw.c [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+	}
+
+#define DW_WDT_INIT(inst)                                                                          \
+	IF_ENABLED(IS_INST_IRQ_EN(inst), (IRQ_CONFIG(inst)))                                       \
+                                                                                                   \
+	static const struct dw_wdt_dev_cfg wdt_dw##inst##_config = {                               \
+		.base = DT_INST_REG_ADDR(inst),                                                    \
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, clock_frequency),                          \
+			(.clk_freq = DT_INST_PROP(inst, clock_frequency)),                         \
+			(.clk_freq = DT_INST_PROP_BY_PHANDLE(inst, clocks, clock_frequency))       \
```
Comment (ceolin) [inline drivers/watchdog/wdt_dw.c:225]: Is this condition is really needed ? I saw that `clock_frequency` is not required property but I couldn't see `clocks` property

Comment (softwarecki) [reply drivers/watchdog/wdt_dw.c:225]: This condition is necessary. The driver requires the `clock_frequency` or `clocks` property. `clocks` is not declared in `snps,designware-watchdog.yaml` because it is included from `base.yaml`. But there is a comment saying that this property is also supported.

File: drivers/watchdog/wdt_dw.c (hunk 10 above)
Comment (softwarecki) [reply drivers/watchdog/wdt_dw.c]: I removed the value check because it is no longer needed since using enum in yaml.

**Review by marcinszkudlinski — APPROVED** [review]

**Review by aborisovich — APPROVED** [review]

**Review by tmleman — APPROVED** [review]
